### PR TITLE
feat(input): add support for more input types

### PR DESCRIPTION
**Pull request by rohit1290**
Fix: https://github.com/Elgg/Elgg/issues/12457

### CODE REVIEW

**Comment by hypeJunction** [comment]
You can pass all those options to input/text

**Comment by rohit1290** [comment]
Well in that case we can pass all options for input/url or input/hidden or input/file using input/text also.

File: views/default/input/color.php [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+ * Displays a color input field
+ *
+ * @package Elgg
+ * @subpackage Core
+ *
+ * Note: type="color" is not supported in:
+ * Internet Explorer 11 or Safari 9.1 (or earlier).
+ */
+
+$vars['class'] = elgg_extract_class($vars, 'elgg-input');
```
Comment (jeabakker) [inline views/default/input/color.php]: the default class should be `elgg-input-color`

File: views/default/input/datetime.php [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ * Displays a datetime-local input field
+ *
+ * @package Elgg
+ * @subpackage Core
+ *
+ * Note: type="datetime-local" is not supported in:
+ * Firefox, Safari or Internet Explorer 12 (or earlier).
+ */
+
+$vars['class'] = elgg_extract_class($vars, 'elgg-input-text');
```
Comment (jeabakker) [inline views/default/input/datetime.php]: the default class should be `elgg-input-datetime`

File: views/default/input/month.php [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+ * Displays a month input field
+ *
+ * @package Elgg
+ * @subpackage Core
+ *
+ * Note: type="month" is not supported in:
+ * Firefox, Safari or Internet Explorer 11 and earlier versions.
+ */
+
+$vars['class'] = elgg_extract_class($vars, 'elgg-input-text');
```
Comment (jeabakker) [inline views/default/input/month.php]: the default class should be `elgg-input-month`

File: views/default/input/search.php [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+/**
+ * Elgg search input
+ * Displays a search input field
+ *
+ * @package Elgg
+ * @subpackage Core
+ *
+ */
+
+$vars['class'] = elgg_extract_class($vars, 'elgg-input-text');
```
Comment (jeabakker) [inline views/default/input/search.php]: the default class should be `elgg-input-search`

File: views/default/input/tel.php [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+/**
+ * Elgg tel input
+ * Displays a tel input field
+ *
+ * @package Elgg
+ * @subpackage Core
+ *
+ */
+
+$vars['class'] = elgg_extract_class($vars, 'elgg-input-text');
```
Comment (jeabakker) [inline views/default/input/tel.php]: the default class should be `elgg-input-tel`

File: views/default/input/tel.php [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+$defaults = [
+	'value' => '',
+	'disabled' => false,
+	'type' => 'tel',
+	'pattern' => '[0-9]{3}-[0-9]{2}-[0-9]{3}'	// Format: 123-45-678
```
Comment (jeabakker) [inline views/default/input/tel.php]: please don't add a default pattern, because it doesn't work for every country. 

Eg. for us in The Netherlands it should be `[0-9]{3,4}-[0-9]{6,7}`

File: views/default/input/week.php [hunk 7]
```diff
@@ -0,0 +1,19 @@
+<?php
+/**
+ * Elgg week input
+ * Displays a week input field
+ *
+ * @package Elgg
+ * @subpackage Core
+ *
+ */
+
```
Comment (jeabakker) [inline views/default/input/week.php]: there is no default class, but it should be `elgg-input-week`

**Review by jeabakker — CHANGES_REQUESTED** [review]
Also please add examples to the theming sandbox of the developers plugin. Maybe you then find that you need to add some CSS.

~Since all other 'text' inputs also have there own files, we should keep that pattern. This also prevents `class` overlapping and therefore styling issues~
Try to get everything to go through `input/text`

Example
```php
/**
 * Elgg color input
 * Displays a color input field
 *
 * Note: type="color" is not supported in:
 * Internet Explorer 11 or Safari 9.1 (or earlier).
 */

$vars['class'] = elgg_extract_class($vars, 'elgg-input-color');

$defaults = [
	'type' => 'color'
];

$vars = array_merge($defaults, $vars);

echo elgg_view('input/text', $vars);
```

The defaults should only set values that matter, no setting empty values.

Also there should be corresponding output views, with correct classes.

**Comment by rohit1290** [comment]
@jeabakker - made all the changes as per your request

File: views/default/input/color.php [hunk 8]
```diff
@@ -0,0 +1,21 @@
+<?php
+/**
+ * Elgg color input
+ * Displays a color input field
+ *
+ * @package Elgg
+ * @subpackage Core
```
Comment (jeabakker) [inline views/default/input/color.php]: please remove all `@package` and `@subpackage` mentions, we don't use them correctly anyway or consistently

File: views/default/elements/forms.css.php [hunk 9]
```diff
@@ -67,7 +67,13 @@
 input[type=password],
 input[type=text],
 input[type=number],
-input[type=url] {
+input[type=url],
+input[type=color],
+input[type=datetime],
```
Comment (jeabakker) [inline views/default/elements/forms.css.php]: this should be `datetime-local`

File: views/default/elements/forms.css.php [hunk 10]
```diff
@@ -76,6 +82,12 @@
 input[type=text]:focus,
 input[type=number]:focus,
 input[type=url]:focus,
+input[type=color]:focus,
+input[type=datetime]:focus,
```
Comment (jeabakker) [inline views/default/elements/forms.css.php]: this should be `datetime-local`

File: views/default/elements/forms.css.php [hunk 11]
```diff
@@ -115,6 +127,21 @@
 	padding-right: 1rem;
 }
 
+.elgg-color-box {
```
Comment (jeabakker) [inline views/default/elements/forms.css.php]: why was this added?

File: views/default/input/month.php [hunk 12]
```diff
@@ -0,0 +1,21 @@
+<?php
+/**
+ * Elgg month input
+ * Displays a month input field
+ *
+ * @package Elgg
+ * @subpackage Core
+ *
+ * Note: type="month" is not supported in:
+ * Firefox, Safari or Internet Explorer 11 and earlier versions.
```
Comment (jeabakker) [inline views/default/input/month.php]: Please remove notes about non supported browsers if there are no version constraints.
You never know if a future version will add support.

**Review by jeabakker — CHANGES_REQUESTED** [review]
some more remarks ;)

Also probably the new `date` fields should support sending it a timestamp

File: views/default/output/color.php [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+$value = elgg_extract('value', $vars);
+
+if (!preg_match('/^#[a-f0-9]{6}$/i', $value)) {
+	return;
+}
+
+$vars['class'] = elgg_extract_class($vars, 'elgg-color-box');
+$vars['style'] = "background-color: $value";
+
+echo elgg_format_element('span', $vars) . " " . elgg_view("output/text", strtoupper($vars));
```
[the remark this answers is not in the document]
Comment (rohit1290) [reply views/default/output/color.php]: the reason I am using strtoupper is because it will be show the hex code in a consistence format. So be it in small letter or capital letter the hex value will always show in upper case.

File: views/default/elements/forms.css.php (hunk 11 above)
Comment (rohit1290) [reply views/default/elements/forms.css.php]: This was added to show the color box for color.php input

Comment (rohit1290) [reply views/default/elements/forms.css.php]: image

**Review by jeabakker — APPROVED** [review]

**Comment by jeabakker** [comment]
Thanks maybe in a follow up PR you could use the new inputs in profile

https://github.com/Elgg/Elgg/blob/73d2af2481e15aea24543950e5c2dbd39e6a8773/mod/profile/start.php#L86-L87

And also add them as an option in the admin defined profile fields

https://github.com/Elgg/Elgg/blob/73d2af2481e15aea24543950e5c2dbd39e6a8773/mod/profile/views/default/forms/profile/fields/add.php#L26-L40
